internal/repos/local.go

```
@@ -0,0 +1,141 @@
+}
+
+func (s *LocalSource) ExternalServices() types.ExternalServices {
+	return types.ExternalServices{s.svc}
+}
+
+func (s *LocalSource) ListRepos(ctx context.Context, results chan SourceResult) {
+	urn := s.svc.URN()
+	repoPaths := getRepoPaths(s.config)
+	for _, r := range repoPaths {
+		uri := "file:///" + r.Path
+		s.logger.Info("found repo ", log.String("uri", uri))
+		results <- SourceResult{
+			Source: s,
+			Repo: &types.Repo{
+				Name: r.fullName(),
+				URI:  uri,
+				ExternalRepo: api.ExternalRepoSpec{
+					ID:          uri,
+					ServiceType: extsvc.TypeLocal,
+					ServiceID:   uri,
+				},
+				Sources: map[string]*types.SourceInfo{
+					urn: {
+						ID:       urn,
+						CloneURL: uri,
+					},
+				},
+				// Looks like this needs to be a JSON object
+				Metadata: "{}",
+			},
+		}
+	}
+}
+
+// Checks if git thinks the given path is a valid .git folder for a repository
+func isBareRepo(path string) bool {
+	c := exec.Command("git", "-C", path, "rev-parse", "--is-bare-repository")
+	out, err := c.CombinedOutput()
+
+	if err != nil {
+		return false
+	}
+
+	return strings.TrimSpace(string(out)) != "false"
+}
+
+// Check if git thinks the given path is a proper git checkout
+func isGitRepo(path string) bool {
+	// Executing git rev-parse --git-dir in the root of a worktree returns .git
+	c := exec.Command("git", "-C", path, "rev-parse")
+	err := c.Run()
+	return err == nil
+}
+
+type repoConfig struct {
+	Path  string
+	Group string
+}
+
+func (c repoConfig) fullName() api.RepoName {
+	name := ""
+	if c.Group != "" {
+		name = c.Group + "/"
+	}
+	name += strings.TrimSuffix(filepath.Base(c.Path), ".git")
+	return api.RepoName(name)
+}
+
+func getRepoPaths(config *schema.LocalExternalService) []repoConfig {
+	paths := []repoConfig{}
+	for _, pathConfig := range config.Repos {
+		pattern, err := homedir.Expand(pathConfig.Pattern)
+		if err != nil {
+			continue
+		}
+		matches, err := filepath.Glob(pattern)
+		if err != nil {
+			continue
```

Do you want silent failures here? Some glob pattern issues seem like can become very hard to debug. Would a warning or log be too noisy here?